sst/SimengCoreWrapper.cc

```diff
@@ -0,0 +1,218 @@
+#include <sst/core/sst_config.h>
+#include "SimengCoreWrapper.hh"
+
+#include <iostream>
+
+// #include "simeng/RegisterValue.hh"
+
+using namespace SST::SSTSimeng;
+using namespace SST::Interfaces;
+// using namespace simeng;
+
+enum class SimulationMode { Emulation, InOrderPipelined, OutOfOrder };
+
+SimengCoreWrapper::SimengCoreWrapper(SST::ComponentId_t id, SST::Params& params): SST::Component(id) {
+    output.init("SimengCoreWrapper[" + getName() + ":@p:@t]:", 999, 0, SST::Output::STDOUT);
+    clock = registerClock(params.find<std::string>("clock", "1GHz"), new SST::Clock::Handler<SimengCoreWrapper>(this, &SimengCoreWrapper::clockTick));
+
+    executable_path = params.find<std::string>("executable_path", "");
+    executable_args = params.find<std::string>("executable_args", "");
+    config_path = params.find<std::string>("config_path", "");
+    cache_line_width  = params.find<uint64_t>("cache_line_width", "64");
+    max_addr_memory = params.find<uint64_t>("max_addr_memory", "0");
+
+    if (executable_path.length() == 0) {
+        output.fatal(CALL_INFO, 10, 0, "Simeng executable binary filepath not provided.");
+    }
+    if (max_addr_memory == 0) {
+      output.fatal(CALL_INFO, 10, 0, "Maximum address range for memory not provided");
+    }
+
+    iterations = 0;
+    vitrual_counter = 0;
+    size = 0;
+
+    mem = loadUserSubComponent<SST::Interfaces::StandardMem>("memory", ComponentInfo::SHARE_NONE, clock,
+      new StandardMem::Handler<SimengCoreWrapper>(this, &SimengCoreWrapper::handleEvent));
+
+    data_memory = std::make_unique<SimengMemInterface>(mem, cache_line_width, max_addr_memory, &output);
+
+    handlers = new SimengMemInterface::SimengMemHandlers(*data_memory, &output);
+
+    registerAsPrimaryComponent();
+    primaryComponentDoNotEndSim();
+}
+
+SimengCoreWrapper::~SimengCoreWrapper() {}
+
+void SimengCoreWrapper::setup() {
+  mem->setup();
+  output.verbose(CALL_INFO, 1, 0, "Memory setup complete\n");
+
+}
+
+void SimengCoreWrapper::handleEvent( StandardMem::Request* ev) {
+  ev->handle(handlers);
+}
+
+void SimengCoreWrapper::finish() {
+
+    output.verbose(CALL_INFO, 1, 0, "Simulation complete. Finalising stats....\n");
+
+    auto endTime = std::chrono::high_resolution_clock::now();
+    auto duration =
+      std::chrono::duration_cast<std::chrono::milliseconds>(endTime - start_time)
+          .count();
+    auto hz = iterations / (static_cast<double>(duration) / 1000.0);
+    auto khz = hz / 1000.0;
+    auto retired = core->getInstructionsRetiredCount();
+    auto mips = retired / static_cast<double>(duration) / 1000.0;
```

Comment:
Would be preferable to decalre the type explicitly instead of using auto for these variables

Reply:
Didn't realise that was the case. I agree with using `auto` for lengthy type names, however these look like `uints` or `doubles`. Maybe something to change in `main` to improve clarity?

Reply:
The time ones might be long and so probably looked nicer to have all as `auto`

Reply:
Shall I leave it as is?

Reply:
Depends on the above comments. Does it look messy/harder to read if auto is not used?

Reply:
PTAL. I have removed `auto` declarations wherever possible.

Reply:
PTAL. I have removed `auto` declarations wherever possible.

Reply:
Which ones?

Reply:
Resolving as it looks fine.